src-electron/rest/user-data.js

```diff
@@ -84,12 +84,123 @@ function httpGetEndpointIds(db) {
  */
 function httpGetDeviceTypeFeatures(db) {
   return async (request, response) => {
-    let deviceTypeRefs = request.query.deviceTypeRefs
-    let deviceTypeFeatures = await queryFeature.getFeaturesByDeviceTypeRefs(
+    let { deviceTypeRefs, endpointTypeRef } = request.query
+    if (Array.isArray(deviceTypeRefs) && deviceTypeRefs.length > 0) {
+      let deviceTypeFeatures = await queryFeature.getFeaturesByDeviceTypeRefs(
+        db,
+        deviceTypeRefs,
+        endpointTypeRef
+      )
+      response.status(StatusCodes.OK).json(deviceTypeFeatures)
+    } else {
+      response.status(StatusCodes.OK).json([])
+    }
+  }
+}
+
+/**
+ * Get all attributes, commands and events in an endpoint type cluster.
+ * @param {*} db
+ * @param {*} endpointTypeClusterId
+ * @param {*} deviceTypeClusterId
+ * @returns elements object containing all attributes, commands and events
+ * in an endpoint type cluster
+ */
+async function getEndpointTypeElements(
+  db,
+  endpointTypeClusterId,
+  deviceTypeClusterId
+) {
+  let elements = {}
+  elements.attributes =
+    await queryAttribute.selectAttributesByEndpointTypeClusterId(
+      db,
+      endpointTypeClusterId
+    )
+  elements.commands =
+    await queryCommand.selectCommandsByEndpointTypeClusterIdAndDeviceTypeClusterId(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+  elements.events =
+    await queryEvent.selectEventsByEndpointTypeClusterIdAndDeviceTypeClusterId(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+  return elements
+}
+
+/**
+ * HTTP POST: elements to be updated after toggle a device type feature
+ *
+ * @param {*} db
+ * @returns callback for the express uri registration
+ */
+function httpPostCheckConformOnFeatureUpdate(db) {
+  return async (request, response) => {
+    let sessionId = request.zapSessionId
+    let { featureData, featureMap, endpointId } = request.body
+    let { endpointTypeClusterId, deviceTypeClusterId } = featureData
+
+    let elements = await getEndpointTypeElements(
+      db,
+      endpointTypeClusterId,
+      deviceTypeClusterId
+    )
+    // check element conform and return elements that need to be updated
+    let result = queryFeature.checkElementConformance(
+      elements,
+      featureMap,
+      featureData,
+      endpointId
+    )
+
+    // set device type feature warning
+    await querySessionNotification.setNotificationOnFeatureChange(
```

Comment:
Same as above: use parallelism.

Reply:
There are dependencies or condition among await calls in this function so I keep it unchanged